src/Sensor/SensorOptical.cpp

```diff
@@ -32,22 +32,27 @@ void SensorOptical::begin() {
 }
 
 void SensorOptical::handle() {
-    #ifdef DEBUG_OPTICAL_ENABLED
-    if (pio_counter_get_count(_pio, 0) != _lastDisplayCount) {
-		DEBUG_SERIAL_LN("pio_counter: " + String(pio_counter_get_count(_pio, _stateMachine)));
-        _lastDisplayCount = pio_counter_get_count(_pio, _stateMachine);
-	}
-    #endif
+    
+    // check if an aperture has passed since the last handle() call
+    if (pio_counter_get_count(_pio, _stateMachine) != _currentCount) {
+        // aperture has passed, so update current count and current aperture time
+        _currentCount = pio_counter_get_count(_pio, _stateMachine);
+        _currentTime = micros();
+
+        
+        #ifdef DEBUG_OPTICAL_ENABLED
+        DEBUG_SERIAL_LN("pio_counter: " + String(pio_counter_get_count(_pio, _stateMachine)));
+        #endif
+    }
 }
 
 float SensorOptical::getAngularVelocity() {
     // n is scaled count apertures we've passed over since last call to getter
-    int32_t count = pio_counter_get_count(_pio, _stateMachine);
-    int32_t n = count - _lastCount;
-    _lastCount = count;
+    int32_t n = _currentCount - _startCount;
+    _startCount = _currentCount;
 
-    uint32_t deltaT = micros() - _lastComputeTime;
-    _lastComputeTime = micros();
+    uint32_t deltaT = _currentTime - _startTime;
```

Comment:
I think the behaviour defined in the second comment is the way to go.  Let me know if you have any questions

Reply:
agreed. 

I wonder if its possible to keep an array of the time in micros for every aperture passage 
```
[micros_1, micros_2, micros_3 ... ]
```
then we can freely call the angular velocity/acceleration function, and we can even specify what resolution we want our function to use. Although one possible concern might be how much memory we have available for the array.

seems like this would also use lazy computation :D

Reply:
I think using an array of micros would be equivalent to taking the average velocity over n apertures, wouldn't it?

Reply:
but there the logic would be quite strange if the code wheel wasn't turning.  I am still for fixed-rate sampling with some form of interpolation, which can be done with graphing software

In general I think it would be good to start with the simplest solution

Reply:
I think the resolution would be something like this

```resolution = 1 / (time period x number of apertures)```

If we take 100 samples per second, the our resolution would be around 100 RPM.

Reply:
I suppose its not too bad when the vehicles are at operating speed (5000-10000 RPM). But with this method there is a tradeoff between time resolution and velocity resolution which I believe can be avoided

Reply:
> My original idea was that we would be able to calculate velocity with a specified resolution. For example, get_velocity(500) would give us the average velocity from the previous 500 milliseconds. Also, having more datapoints will make acceleration calculations more accurate. Of course, just taking the average velocity n apertures would be much simpler and I would support that as well.

So you you want to be able to select from a set of moving averages?  We could do that without an array of values.  What advantage would this have over simply having SensorOptical return its current best approximation of angular velocity?

> I think the resolution would be something like this
resolution = 1 / (time period x number of apertures)
If we take 100 samples per second, the our resolution would be around 100 RPM.

If we have samples like: { 0, 0, 1/(tn), 0, 0, 1/(tn), ... } the moving average over these 6 samples would be 1/(3tn), which is fairly accurate.  This could be computed while data is being recorded, or left as is so you can play around with different moving averages when you graph it out

Reply:
ahh okay, i understand what you mean now

Reply:
I wonder what the best way to receive data from SensorOptical would be. If the member fields for velocity/acceleration are being updated in the handle method, if we don't want to lose any data, we would need to know when they are being updated so we can record their changes (or we could take a sample every time we call the handle method).

Reply:
I wonder if we can create a separate data manager class, that can drive the variable updates in the SensorOptical class + save all the samples.

Reply:
Since SensorOptical::handle is going to be called from the main control loop, and whatever data-logging methods we are using is also going to be happening from the control loop, we could have logging and averaging on the same interval.

In that case it would be useful to have a method: SensorOptical::getWindowSize (maybe there's a better name), if you want the average window size to be controllable, then you can add a setter

Reply:
> Since SensorOptical::handle is going to be called from the main control loop, and whatever data-logging methods we are using is also going to be happening from the control loop, we could have logging and averaging on the same interval.

This is why I set it up to compute angular velocity lazily in the first place

Reply:
I will work on the code for averaging + logging the data during this week

Reply:
I was trying to suggest that computing averages be done with graphing software.  The whole idea of using csv files is so you can take the raw data and graph it out using something more sophisticated, no?

@underscoreHasan is currently working on logging data, but you can have a look in this project's issues panel if you're looking for other things to do.

Reply:
that is also a very short interval for computing angular velocity.  Do you know you need to read that often?  The speed of the dyno rollers changes quite slowly

Reply:
merging it now.  I made some changes:
- fixed a compiler error (there was a reference to an field that had been renamed
- angular velocity and linear velocity were implicitly being cast from double to float
- I added a getter and setter for SensorOptical::_readInterval